.changeset/lucky-carrots-sin.md

```diff
@@ -0,0 +1,28 @@
+---
+"uploadthing": minor
+"@uploadthing/shared": minor
+---
+
+feat: add ability to provide custom identifiers when uploading files
+
+also adds ability for UTApi methods to filter based on custom identifiers
+
+```ts
+// bind custom id when uploading
+f(["image"])
+  .middleware(({ files }) => {
+    const filesWithIds = files.map((f) => ({
```

Comment:
IMO this is fine for now, but we should rethink the structure of this builder pattern in the future. Likely needs to be multi-step, i.e. "authenticate", "preUpload" and "onUploadComplete"